drift/lib/src/query_builder/statements/delete.dart

```diff
@@ -17,25 +22,53 @@ final class DeleteStatement<Row extends Object,
   final GeneratedTable<Row, RS> resultSet;
 
   /// An optional `RETURNING` clause part of this statement.
-  ReturningClause<Row, RS>? returning;
+  final ReturningClause<Row, RS>? returning;
 
   final DatabaseConnectionUser _database;
 
+  @override
+  final WhereClause? whereClause;
+
+  @override
+  final BuiltMap<Symbol, Object?> dialectSpecificOptions = BuiltMap();
+
   /// This constructor should be called by [DatabaseConnectionUser.delete] for
   /// you.
-  DeleteStatement(this._database, this.resultSet);
+  DeleteStatement(this._database, this.resultSet)
+      : returning = null,
+        whereClause = null;
+  DeleteStatement._(this._database,
+      {required this.resultSet,
+      required this.whereClause,
+      required this.returning});
+
+  @override
+  DeleteStatement<Row, RS> withWhereClause(WhereClause whereClause) =>
+      _copyWith(whereClause: whereClause);
+
+  DeleteStatement<Row, RS> _withReturning() =>
+      _copyWith(returning: ReturningClause(resultSet));
+
+  DeleteStatement<Row, RS> _copyWith({
+    GeneratedTable<Row, RS>? resultSet,
+    WhereClause? whereClause,
```

Comment:
The problem with these kinds of `copyWith` methods is that you can't set fields back to `null`.

This is why I would suggest an approach where:

1. There is no public API to mutate statements (so no mutable `WhereClause? whereClause;`).
2. But, `WhereClause?` is a mutable field internally, encapsulated through a getter.
3. Then, when we want to add a where clause, we do something like `DeleteStatement<Row, RS> where(Expression<bool> expression) => copy().._whereClause = ...`.

Reply:
There are very few fields which are mutated. The solution for this is to create a `copyWith*` fields for each field. This would allow you to set a value back to null.
```dart
class Foo {
  final int bar;
  Foo copyWithBar(int? newBar){
    return Foo(newBar);
  }
}
```
There are not _that_ many fields, it should not be too big of an issue.

Reply:
N/A